core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -259,20 +261,21 @@ class Partition(val topic: String,
   }
 
   /**
-   * Check and maybe expand the ISR of the partition.
+   * Check and maybe expand the ISR of the partition. A replica is in ISR of the partition if and only if
+   * replica's LEO >= HW and replica's lag <= replicaLagTimeMaxMs.
    *
    * This function can be triggered when a replica's LEO has incremented
    */
-  def maybeExpandIsr(replicaId: Int) {
+  def maybeExpandIsr(replicaId: Int, logReadResult: LogReadResult) {
     val leaderHWIncremented = inWriteLock(leaderIsrUpdateLock) {
       // check if this replica needs to be added to the ISR
       leaderReplicaIfLocal() match {
         case Some(leaderReplica) =>
           val replica = getReplica(replicaId).get
-          val leaderHW = leaderReplica.highWatermark
           if(!inSyncReplicas.contains(replica) &&
              assignedReplicas.map(_.brokerId).contains(replicaId) &&
-                  replica.logEndOffset.offsetDiff(leaderHW) >= 0) {
+             logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.hw &&
+             logReadResult.fetchTimeMs - replica.lastCaughtUpTimeMs <= replicaManager.config.replicaLagTimeMaxMs) {
```

Comment:
Sorry, my explanation above is wrong. Let me correct it.

I think it is still needed. Otherwise a follower will be added to ISR even if its LEO < HW. This is problematic because a message with offset = HW is not necessarily persisted by all replicas in the ISR.

For example, let's say replication factor = 2, isr includes only the leader. At time t1, hw = 100 and leader's LEO = 100. The leader receives follower's fetch request at t1 with offset = 90, and sends back fetch response in offset range [90, 100). At time t2, hw = 120 and leader's LEO = 120. The leader receives follower's fetch request at t2 with offset = 100. Suppose t2 - t1 < replicaLagTimeMaxMs, if we don't have 2nd check, the follower would be added to the isr set and even if its LEO is smaller than hw.